### PR TITLE
[Feature] Add Fast Visibility Graph Based ECG R-peak Detector

### PR DESCRIPTION
**Pull request by JonasEmrich**
Adding the accelerated visbility-graph-based R-peak detector (https://ieeexplore.ieee.org/document/10290007, https://github.com/JonasEmrich/vg-beat-detectors) #938

### CODE REVIEW

File: neurokit2/ecg/ecg_findpeaks.py [hunk 1]
```diff
@@ -105,12 +97,12 @@ def _ecg_findpeaks_findmethod(method):
         return _ecg_findpeaks_rodrigues
     elif method in ["vg", "vgraph", "koka2022"]:
         return _ecg_findpeaks_vgraph
+    elif method in ["FastNVG", "fastnvg", "emrich", "emrich2023"]:
```
Comment (DominiqueMakowski) [inline neurokit2/ecg/ecg_findpeaks.py:100]: `method = method.lower()` above so you don't need the upper case variant :)

Comment (JonasEmrich) [reply neurokit2/ecg/ecg_findpeaks.py]: True. I have fixed it in [7674b25](https://github.com/neuropsychology/NeuroKit/pull/939/commits/7674b25f5b3d5d44f4c6cad82c77f3531d9c9588)

**Comment by JonasEmrich** [comment]
Now, I should have adapted the package at all the required places. Together with the authors of the koka2022 method, we decided to remove it in favor of this more advanced method. 

I tested the implementation briefly on two databases, and it appears that it procures results comparable to the original FastNVG algorithm (https://github.com/JonasEmrich/vg-beat-detectors)

**Comment by DominiqueMakowski** [comment]
Amazing job! Thanks

> Together with the authors of the koka2022 method, we decided to remove it in favor of this more advanced method.

In that case, could we leave the `"koka2022"` input to then run your udpated method, throwing a warning that `"The 'koka2022' method has been replaced by XX. Please replace method='koka2022' by ..."` (so that we don't break potential user code just yet)

**Comment by JonasEmrich** [comment]
> In that case, could we leave the `"koka2022"` input to then run your udpated method, throwing a warning that `"The 'koka2022' method has been replaced by XX. Please replace method='koka2022' by ..."` (so that we don't break potential user code just yet)

Thanks for pointing this out! I added a warning in `ecg_clean()` and `ecg_findpeaks()`.

**Comment by DominiqueMakowski** [comment]
I think we can just add it in `ecg_findpeaks()` because it's there that the switch happens

**Comment by JonasEmrich** [comment]
> I think we can just add it in `ecg_findpeaks()` because it's there that the switch happens

Although the emrich2023 method uses the same cleaning as koka2022 (points to the same function), I think that perhaps users also should use emrich2023 here instead of koka2022. It's just the naming but makes it more consistent, since the koka2022 method is deprecated/removed. Or what are your thoughts?

**Comment by DominiqueMakowski** [comment]
> It's just the naming but makes it more consistent, since the koka2022 method is deprecated/removed

makes sense!

**Comment by DominiqueMakowski** [comment]
@JonasEmrich is it good for me to merge?

**Comment by JonasEmrich** [comment]
> @JonasEmrich is it good for me to merge?

@DominiqueMakowski yes, from my side it is ready to merge

**Comment by DominiqueMakowski** [comment]
Amazing, thanks again for all the work :)

**Comment by NikkiNgombe** [comment]
Hello @JonasEmrich and @DominiqueMakowski ,
I'm working with some ECG and wanted to use "emrich2023" for the peak detection. However, I get the following error:

ValueError: NeuroKit error: ecg_findpeaks(): 'emrich2023' not implemented.

After installing ts2vg, I was able to use koka2022, but I am still unable to run ecg_findpeaks() using the Emrich 2023 method. Is this emrich2023 method actually implemented in NeuroKit2? When I checked the source code for ecg_findpeaks(), I also noticed, that emrich2023 wasn't listed in the available methods.
Screenshot 2024-07-02 at 20 30 42

Any help would be greatly appreciated, thank you!
Nikki

**Comment by DominiqueMakowski** [comment]
I can see the lines in the current version: https://github.com/neuropsychology/NeuroKit/blob/1aa8deee392f8098df4fd77a23f696c2ff2d29db/neurokit2/ecg/ecg_clean.py#L126-L127

Do you have the latest version of NK installed?

**Comment by NikkiNgombe** [comment]
Thank you for the quick response! I was using version 0.2.7. I have now updated to version 0.2.9 and can now use the emrich2023 option. Thank you so much! :)

**Pull request merged**